README.md
# i-doit Docker images

Dockerized [i-doit CMDB](https://i-doit.com/)

[![Stars](https://img.shields.io/docker/stars/bheisig/idoit.svg)](https://hub.docker.com/r/bheisig/idoit/)
[![Pulls](https://img.shields.io/docker/pulls/bheisig/idoit.svg)](https://hub.docker.com/r/bheisig/idoit/)
[![Image size](https://img.shields.io/docker/image-size/bheisig/idoit?sort=date)](https://hub.docker.com/r/bheisig/idoit/)
![Build status](https://github.com/i-doit/docs/actions/workflows/main.yml/badge.svg?branch=main)

## Not production-ready!

**The provided Docker images aren't production-ready. Please do not rely on them. They come "as is" with no warranty of any kind. They are not an official product by synetics GmbH. Because of this, we can't provide you any support.**

At the moment, we use these images for quality tests of i-doit and its add-ons. In the near-future we hope to provide you official i-doit Docker images which are well-tested and production-ready. In the meantime don't hesitate to test these images yourself. [We love to get your feedback!](https://github.com/bheisig/i-doit-docker/issues)

## Supported tags and respective `Dockerfile` links

-   `1.19-open-php8.0-fpm` ([`Dockerfile`](1.19/open/php8.0/fpm))
-   `1.19-open-php8.0-apache` ([`Dockerfile`](1.19/open/php8.0/apache))
-   `1.19-pro-php8.0-fpm` ([`Dockerfile`](1.19/pro/php8.0/fpm))
-   `1.19-pro-php8.0-apache` ([`Dockerfile`](1.19/pro/php8.0/apache))
-   `1.19-open-php7.4-fpm` ([`Dockerfile`](1.19/open/php7.4/fpm))
-   `1.19-open-php7.4-apache` ([`Dockerfile`](1.19/open/php7.4/apache))
-   `1.19-pro-php7.4-fpm` ([`Dockerfile`](1.19/pro/php7.4/fpm))
-   `1.19-pro-php7.4-apache` ([`Dockerfile`](1.19/pro/php7.4/apache))
-   `1.18.2-open-php7.4-fpm` ([`Dockerfile`](1.18.2/open/php7.4/fpm))
-   `1.18.2-open-php7.4-apache` ([`Dockerfile`](1.18.2/open/php7.4/apache))
-   `1.18.2-pro-php7.4-fpm` ([`Dockerfile`](1.18.2/pro/php7.4/fpm))
-   `1.18.2-pro-php7.4-apache` ([`Dockerfile`](1.18.2/pro/php7.4/apache))
-   `1.18.1-open-php7.4-fpm` ([`Dockerfile`](1.18.1/open/php7.4/fpm))
-   `1.18.1-open-php7.4-apache` ([`Dockerfile`](1.18.1/open/php7.4/apache))
-   `1.18.1-pro-php7.4-fpm` ([`Dockerfile`](1.18.1/pro/php7.4/fpm))
-   `1.18.1-pro-php7.4-apache` ([`Dockerfile`](1.18.1/pro/php7.4/apache))
-   `1.18-open-php7.4-fpm` ([`Dockerfile`](1.18/open/php7.4/fpm))
-   `1.18-open-php7.4-apache` ([`Dockerfile`](1.18/open/php7.4/apache))
-   `1.18-pro-php7.4-fpm` ([`Dockerfile`](1.18/pro/php7.4/fpm))
-   `1.18-pro-php7.4-apache` ([`Dockerfile`](1.18/pro/php7.4/apache))

**Note:** Older branches are not supported anymore but can still be found in the repository.

## Quick reference

-   **Where to get help**:
    [`https://github.com/bheisig/i-doit-docker/issues`](https://github.com/bheisig/i-doit-docker/issues)
-   **Where to file issues**:
    [`https://github.com/bheisig/i-doit-docker/issues`](https://github.com/bheisig/i-doit-docker/issues)
-   **Maintained by**:
    [Benjamin Heisig](https://benjamin.heisig.name/)
-   **Supported architectures**: x86_64
-   **Source of this description**:
    [`README.md`](https://github.com/bheisig/i-doit-docker/blob/master/README.md)
-   **Supported Docker versions**:
    [the latest release](https://github.com/docker/docker-ce/releases/latest) (down to 1.6 on a best-effort basis)

## What is i-doit?

i-doit ("I document IT") is a fully-featured Web application for CMDB (Configuration Management Databases), IT asset management (ITAM) and IT documentation. i-doit is maintained by the company synetics GmbH, located in Düsseldorf/Germany.

## How to use this image

Each image is built on the official PHP Docker image. We've got heavily inspired by the well-documented [Nextcloud docker image](https://github.com/nextcloud/docker). Thank you!

### Run i-doit with PHP-FPM (recommended)

Run the latest version of i-doit open with PHP 8.0 and FPM:

~~~ {.bash}
docker run --name i-doit-fpm -p 9000:9000 bheisig/idoit:1.19-open-php8.0-fpm
~~~

### Run i-doit with Apache HTTPD (legacy)

Run the latest version of i-doit open with PHP 8.0 and Apache HTTPD 2.4:

~~~ {.bash}
docker run --name i-doit-apache -p 80:80 bheisig/idoit:1.19-open-php8.0-apache
~~~

### Available volumes

Each i-doit container has one default volume containing the complete installation directory: `/var/www/html`. This directory includes the source code, cache files, uploaded files, installed add-ons, custom translation files, etc.

### Run i-doit with Docker Compose

i-doit requires either MariaDB or MySQL as the database backend. For better performance Memcached is highly recommended. There are some examples:

-   [Run with PHP-FPM](docker-compose-fpm.yml) (preferred)
-   [Run with PHP-FPM over UNIX sockets](docker-compose-sockets.yml) (even faster)
-   [Run with Apache HTTPD](docker-compose-apache.yml) (legacy)

### Run i-doit CLI tool

i-doit has its own CLI tool named `console.php` for long-lasting, recurring tasks in background. You can call any CLI command in the running Docker container, for example:

~~~ {.bash}
docker exec -it --user www-data i-doit-fpm php console.php --help
~~~

### Update i-doit

_to be defined_

### Backup and restore

For a complete backup and restore process you need to consider at least 3 sources:

1.  i-doit installation directory (see section "Available volumes")
2.  System database (default: `idoit_system`)
3.  Each tenant database (default 1st one: `idoit_data`)

As you can see in the examples for `docker-compose` (see section "Run i-doit with Docker Compose") the simplest thing is to backup the named volumes. But for a running instance of MariaDB this is a bad solution because everything stored temporarily in memory won't be backed up. Run `mysqldump` (for backup) and `mysql` (for restore) to fetch every bit stored in the databases.

This is a basic example to backup everything in a running environment:

~~~ {.bash}
docker exec i-doit-fpm /bin/tar cvf - . | gzip -9 > backup.tar.gz
docker exec i-doit-fpm /usr/bin/mysqldump -uidoit -pidoit --all-databases | gzip -9 > backup.sql.gz
~~~

This is a basic example to restore those backups:

~~~ {.bash}
cat backup.tar.gz | docker exec --interactive --user www-data i-doit-fpm /bin/tar xzvf -
gunzip < backup.sql.gz | docker exec --interactive i-doit-fpm /usr/bin/mysql -uidoit -pidoit
~~~

Don't forget to alter the commands above to your needs.

### Add a subscription license

Copy the file to the running i-doit container and import the file with the i-doit CLI tool:

~~~ {.bash}
docker cp license.txt i-doit-fpm:/tmp/
docker exec --interactive --user www-data i-doit-fpm php \
    console.php license-add \
    --user admin --password admin --no-interaction --license /tmp/license.txt
~~~

### Move your CMDB data to a container

The next example assumes that i-doit is installed on the same host as the running i-doit container:

~~~ {.bash}
cd /var/www/html/
tar czvf - . | docker exec --interactive --user www-data i-doit-fpm /bin/tar xzvf -
mysqldump -uidoit -pidoit --all-databases | \
    docker exec --interactive i-doit-fpm /usr/bin/mysql -uidoit -pidoit
~~~

### TLS/HTTPS

We _strongly_ recommend to access i-doit in a productive environment via TLS/HTTPS. Consider to use a load balancer or reverse proxy because the provided images are unable to handle TLS/HTTPS connections themselves.

## Image variants

`<version>-<edition>-<php>-<service>`

-   `<version>`: Which version of i-doit do you like to run?
-   `<edition>`: Which edition of i-doit do you like to run? Select `open` for community edition (Free Software) and `pro` for professional edition (proprietary).
-   `<php>`: Which version of [PHP](https://php.net/) do you like to run?
-   `<service>`: Decide to run i-doit either with [Apache HTTPD](https://httpd.apache.org/) or [PHP-FPM](https://php-fpm.org/)

## Copyright & license

Copyright (C) 2019-2022 [synetics GmbH](https://i-doit.com/)

Licensed under the [GNU Affero GPL version 3 or later (`AGPLv3+`)](https://gnu.org/licenses/agpl.html)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased][]

### Added

-   Add support of i-doit open/pro, versions `1.18.2` and `1.19`
-   Add support of PHP version branch `8.0`

### Changed

-   Re-factor setup routine

### Removed

-   Drop support of i-doit open/pro `1.17` branch

## [0.5.0][] – 2022-04-25

### Added

-   Add support of i-doit open/pro, versions `1.17.2`, `1.18` and `1.18.1`

### Removed

-   Drop support of i-doit open/pro `1.16` branch

## [0.4.0][] – 2021-10-27

### Added

-   Add support of i-doit open, versions `1.16`, `1.16.1`, `1.16.2`, `1.16.3`, `1.17`, `1.17.1`
-   Add support of i-doit pro, versions `1.16`, `1.16.1`, `1.16.2`, `1.16.3`, `1.17`, `1.17.1`

### Changed

-   Bump version of MariaDB; recommended version is `v10.5.x`
-   Upgrade base container image Debian GNU/Linux from 10 `buster` to 11 `bullseye`

### Removed

-   Drop support of i-doit open/pro `1.14` and `1.15` branches
-   Drop support of PHP `7.1`, `7.2`, and `7.3` (end of life)
-   Remove command `./docker.sh test` in favor of `npm test`

## [0.3.0][] – 2020-10-23

### Added

-   Add i-doit open, version `1.15.1`
-   Add i-doit pro, version `1.15.1`
-   Add i-doit open, version `1.15`
-   Add i-doit pro, version `1.15`
-   Add i-doit open, version `1.14.2`

### Fixed

-   `apache` service: Add content from i-doit's `.htaccess` file to Apache's configuration file

### Removed

-   Do not build images of `1.13.x` branch anymore

## [0.2.0][] – 2020-06-02

### Added

-   Add i-doit pro/open, versions `1.13.1`, `1.13.2`, `1.14`, `1.14.1` and `1.14.2`
-   Scan images for vulnerabilities with `./docker.sh scan`
-   Add `./docker.sh fix` to fix file permissions

### Fixed

-   Ensure to remove all remaining docker images while cleaning up

### Removed

-   Do not build images of `1.12.x` branch anymore

## 0.1.0 – 2019-07-24

Initial release

### Added

-   Add i-doit versions `1.12.1`, `1.12.4` and `1.13`
-   Add i-doit editions open and pro
-   Add PHP versions `7.0`, `7.1`, `7.2`, `7.3` and `7.4` (experimental)
-   Add services Apache Web server and PHP-FPM
-   Add examples for docker-compose incl. environment variables, MariaDB and Memcached

[Unreleased]: https://github.com/bheisig/i-doit-docker/compare/v0.5.0...HEAD
[0.5.0]: https://github.com/bheisig/i-doit-docker/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/bheisig/i-doit-docker/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/bheisig/i-doit-docker/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/bheisig/i-doit-docker/compare/v0.1.0...v0.2.0


docker.sh
#!/bin/bash

set -euo pipefail
IFS=$'\n\t'

: "${DOCKER_USERNAME:=""}"
: "${DOCKER_PASSWORD:=""}"
: "${DOCKER_IMAGE:="bheisig/idoit"}"

function execute {
    local job="$1"

    case "$job" in
        "build")
            buildImages
            ;;
        "clean")
            cleanUp
            ;;
        "fix")
            fix
            ;;
        "help")
            printUsage
            ;;
        "login")
            loginToDockerHub
            ;;
        "logout")
            logoutFromDockerHub
            ;;
        "print")
            printReadme
            ;;
        "pull")
            pullImages
            ;;
        "push")
            pushImages
            ;;
        "scan")
            scanImages
            ;;
        *)
            printUsage
            ;;
    esac
}

function loginToDockerHub {
    test -n "$DOCKER_USERNAME" || \
        abort "Unknown Docker ID"

    test -n "$DOCKER_PASSWORD" || \
        abort "Empty password for Docker ID $DOCKER_USERNAME"

    echo "$DOCKER_PASSWORD" | \
        docker login -u "$DOCKER_USERNAME" --password-stdin || \
        abort "Unable to sign in to Docker Hub with Docker ID $DOCKER_USERNAME"
}

function logoutFromDockerHub {
    docker logout || \
    abort "Unable to logout from Docker Hub"
}

function buildImages {
    log "Build Docker images…"

    buildImage 1.18 pro php7.4 apache
    buildImage 1.18 pro php7.4 fpm

    buildImage 1.18 open php7.4 apache
    buildImage 1.18 open php7.4 fpm

    buildImage 1.18.1 pro php7.4 apache
    buildImage 1.18.1 pro php7.4 fpm

    buildImage 1.18.1 open php7.4 apache
    buildImage 1.18.1 open php7.4 fpm

    buildImage 1.18.2 pro php7.4 apache
    buildImage 1.18.2 pro php7.4 fpm

    buildImage 1.18.2 open php7.4 apache
    buildImage 1.18.2 open php7.4 fpm

    buildImage 1.19 pro php7.4 apache
    buildImage 1.19 pro php7.4 fpm

    buildImage 1.19 open php7.4 apache
    buildImage 1.19 open php7.4 fpm

    buildImage 1.19 pro php8.0 apache
    buildImage 1.19 pro php8.0 fpm

    buildImage 1.19 open php8.0 apache
    buildImage 1.19 open php8.0 fpm
}

function pullImages {
    log "Pull Docker images from registry…"

    pullImage php:8.0-fpm-bullseye
    pullImage php:8.0-apache-bullseye
    pullImage php:7.4-fpm-bullseye
    pullImage php:7.4-apache-bullseye
}

function pullImage {
    local image="$1"

    log "Pull $image from Docker registry"

    docker pull "$image" || \
        abort "No pull"
}

function buildImage {
    local version="$1"
    local edition="$2"
    local php="$3"
    local service="$4"
    local path="${version}/${edition}/${php}/${service}/"
    local tag="${DOCKER_IMAGE}:${version}-${edition}-${php}-${service}"

    log "Build $tag from $path"

    docker build \
        -t "$tag" \
        "$path" || \
        abort "No build"
}

function scanImages {
    log "Scan Docker images for vulnerabilities…"

    scanImage 1.18 pro php7.4 apache
    scanImage 1.18 pro php7.4 fpm

    scanImage 1.18 open php7.4 apache
    scanImage 1.18 open php7.4 fpm

    scanImage 1.18.1 pro php7.4 apache
    scanImage 1.18.1 pro php7.4 fpm

    scanImage 1.18.1 open php7.4 apache
    scanImage 1.18.1 open php7.4 fpm

    scanImage 1.18.2 pro php7.4 apache
    scanImage 1.18.2 pro php7.4 fpm

    scanImage 1.18.2 open php7.4 apache
    scanImage 1.18.2 open php7.4 fpm

    scanImage 1.19 pro php7.4 apache
    scanImage 1.19 pro php7.4 fpm

    scanImage 1.19 open php7.4 apache
    scanImage 1.19 open php7.4 fpm

    scanImage 1.19 pro php8.0 apache
    scanImage 1.19 pro php8.0 fpm

    scanImage 1.19 open php8.0 apache
    scanImage 1.19 open php8.0 fpm
}

function scanImage {
    local version="$1"
    local edition="$2"
    local php="$3"
    local service="$4"
    local tag="${DOCKER_IMAGE}:${version}-${edition}-${php}-${service}"

    log "Scan image $tag"
    trivy --quiet image --severity HIGH,CRITICAL "$tag"
}

function fix {
    fixFilePermissions
}

function fixFilePermissions {
    find . -type f -iname "*.sh" -exec chmod +x {} \; || \
        abort "Not good"
}

function pushImages {
    log "Push Docker images to registry…"

    pushImage 1.18 pro php7.4 apache
    pushImage 1.18 pro php7.4 fpm

    pushImage 1.18 open php7.4 apache
    pushImage 1.18 open php7.4 fpm

    pushImage 1.18.1 pro php7.4 apache
    pushImage 1.18.1 pro php7.4 fpm

    pushImage 1.18.1 open php7.4 apache
    pushImage 1.18.1 open php7.4 fpm

    pushImage 1.18.2 pro php7.4 apache
    pushImage 1.18.2 pro php7.4 fpm

    pushImage 1.18.2 open php7.4 apache
    pushImage 1.18.2 open php7.4 fpm

    pushImage 1.19 pro php7.4 apache
    pushImage 1.19 pro php7.4 fpm

    pushImage 1.19 open php7.4 apache
    pushImage 1.19 open php7.4 fpm

    pushImage 1.19 pro php8.0 apache
    pushImage 1.19 pro php8.0 fpm

    pushImage 1.19 open php8.0 apache
    pushImage 1.19 open php8.0 fpm
}

function pushImage {
    local version="$1"
    local edition="$2"
    local php="$3"
    local service="$4"
    local tag="${DOCKER_IMAGE}:${version}-${edition}-${php}-${service}"

    log "Push image $tag to Docker registry"

    docker push "$tag" || \
        abort "No push, no forward"
}

function printReadme {
    printSupportedTags 1.19 open php8.0 fpm
    printSupportedTags 1.19 open php8.0 apache

    printSupportedTags 1.19 pro php8.0 fpm
    printSupportedTags 1.19 pro php8.0 apache

    printSupportedTags 1.19 open php7.4 fpm
    printSupportedTags 1.19 open php7.4 apache

    printSupportedTags 1.19 pro php7.4 fpm
    printSupportedTags 1.19 pro php7.4 apache

    printSupportedTags 1.18.2 open php7.4 fpm
    printSupportedTags 1.18.2 open php7.4 apache

    printSupportedTags 1.18.2 pro php7.4 fpm
    printSupportedTags 1.18.2 pro php7.4 apache

    printSupportedTags 1.18.1 open php7.4 fpm
    printSupportedTags 1.18.1 open php7.4 apache

    printSupportedTags 1.18.1 pro php7.4 fpm
    printSupportedTags 1.18.1 pro php7.4 apache

    printSupportedTags 1.18 open php7.4 fpm
    printSupportedTags 1.18 open php7.4 apache

    printSupportedTags 1.18 pro php7.4 fpm
    printSupportedTags 1.18 pro php7.4 apache
}

function printSupportedTags {
    local version="$1"
    local edition="$2"
    local php="$3"
    local service="$4"
    local tag=""
    local path=""

    tag="${version}-${edition}-${php}-${service}"
    path="${version}/${edition}/${php}/${service}"

    log "-   \`${tag}\` ([\`Dockerfile\`](${path}))"
}

function cleanUp {
    local amount="0"

    log "Clean up"

    amount="$(docker images -q "${DOCKER_IMAGE}" | wc -l)"

    case "$amount" in
        "0")
            log "Already cleaned up"
            return 0
            ;;
        "1")
            log "Remove 1 docker image by name ${DOCKER_IMAGE}"
            ;;
        *)  log "Remove $amount docker images by name ${DOCKER_IMAGE}"
            ;;
    esac

    # We need word splitting:
    # shellcheck disable=SC2046
    docker rmi -f $(docker images -q "${DOCKER_IMAGE}") || \
        abort "Unable to remove docker images"
}

function printUsage {
    log "build|clean|fix|help|login|logout|print|pull|push|scan"
}

function setUp {
    test "$(whoami)" != root || \
        log "Please do not run this script as root user"

    command -v docker > /dev/null || \
        abort "Command \"docker\" is missing"
}

function finish {
    exit 0
}

function abort {
    log "$1"
    exit 1
}

function log {
    echo -e "$1"
}

if [[ "${BASH_SOURCE[0]}" = "$0" ]]; then
    setUp && execute "${1-help}" && finish
fi
